File: CommonBusiness/src/main/java/com/company/commonbusiness/util/ActivityUtils.java
package com.company.commonbusiness.util;

import android.app.Activity;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.content.pm.ResolveInfo;
import android.os.Build;
import android.support.annotation.NonNull;
import android.util.ArrayMap;


import com.company.commonbusiness.base.activity.BaseActivity;

import java.lang.reflect.Field;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * @author 李昭鸿
 * @desc: Activity相关工具类
 * @date Created on 2017/7/21 17:14
 */

public final class ActivityUtils {

    private static List<BaseActivity> activities = new ArrayList<>();

    private ActivityUtils() {
        throw new UnsupportedOperationException("u can't instantiate me...");
    }

    /**
     * 添加Activity到列表
     *
     * @param activity 添加的activity
     */
    public static void addActivity(BaseActivity activity) {
        activities.add(activity);
    }

    /**
     * 从列表中移除Activity
     *
     * @param activity 移除的activity
     */
    public static void removeActivity(BaseActivity activity) {
        activities.remove(activity);
    }

    /**
     * 用于退出app时结束所有Activity
     */
    public static void finishAll() {
        for (BaseActivity activity : activities) {
            if (!activity.isFinishing()) {
                activity.finish();
            }
        }
    }

    /**
     * 判断是否存在Activity
     *
     * @param packageName 包名
     * @param className   activity全路径类名
     * @return {@code true}: 是<br>{@code false}: 否
     */
    public static boolean isActivityExists(@NonNull final String packageName, @NonNull final String className) {
        Intent intent = new Intent();
        intent.setClassName(packageName, className);
        return !(Utils.getContext().getPackageManager().resolveActivity(intent, 0) == null ||
                intent.resolveActivity(Utils.getContext().getPackageManager()) == null ||
                Utils.getContext().getPackageManager().queryIntentActivities(intent, 0).size() == 0);
    }

    /**
     * 获取launcher activity
     *
     * @param packageName 包名
     * @return launcher activity
     */
    public static String getLauncherActivity(@NonNull final String packageName) {
        Intent intent = new Intent(Intent.ACTION_MAIN, null);
        intent.addCategory(Intent.CATEGORY_LAUNCHER);
        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        PackageManager pm = Utils.getContext().getPackageManager();
        List<ResolveInfo> info = pm.queryIntentActivities(intent, 0);
        for (ResolveInfo aInfo : info) {
            if (aInfo.activityInfo.packageName.equals(packageName)) {
                return aInfo.activityInfo.name;
            }
        }
        return "no " + packageName;
    }

    /**
     * 获取栈顶Activity
     *
     * @return 栈顶Activity
     */
    public static Activity getTopActivity() {
        try {
            Class activityThreadClass = Class.forName("android.app.ActivityThread");
            Object activityThread = activityThreadClass.getMethod("currentActivityThread").invoke(null);
            Field activitiesField = activityThreadClass.getDeclaredField("mActivities");
            activitiesField.setAccessible(true);
            Map activities;
            if (Build.VERSION.SDK_INT < Build.VERSION_CODES.KITKAT) {
                activities = (HashMap) activitiesField.get(activityThread);
            } else {
                activities = (ArrayMap) activitiesField.get(activityThread);
            }
            for (Object activityRecord : activities.values()) {
                Class activityRecordClass = activityRecord.getClass();
                Field pausedField = activityRecordClass.getDeclaredField("paused");
                pausedField.setAccessible(true);
                if (!pausedField.getBoolean(activityRecord)) {
                    Field activityField = activityRecordClass.getDeclaredField("activity");
                    activityField.setAccessible(true);
                    return (Activity) activityField.get(activityRecord);
                }
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
        return null;
    }
}

File: CommonBusiness/src/main/java/com/company/commonbusiness/util/ActivityUtils.kt
package com.company.commonbusiness.util

import android.app.Activity
import android.content.Intent
import android.content.pm.PackageManager
import android.content.pm.ResolveInfo
import android.os.Build
import android.util.ArrayMap


import com.company.commonbusiness.base.activity.BaseActivity

import java.lang.reflect.Field
import java.util.ArrayList
import java.util.HashMap

/**
 * @author 李昭鸿
 * @desc: Activity相关工具类
 * @date Created on 2017/7/21 17:14
 */

class ActivityUtils private constructor() {

    init {
//        throw UnsupportedOperationException("u can't instantiate me...")
    }

    companion object {

        private val activities = ArrayList<BaseActivity>()

        /**
         * 添加Activity到列表
         *
         * @param activity 添加的activity
         */
        fun addActivity(activity: BaseActivity) {
            activities.add(activity)
        }

        /**
         * 从列表中移除Activity
         *
         * @param activity 移除的activity
         */
        fun removeActivity(activity: BaseActivity) {
            activities.remove(activity)
        }

        /**
         * 用于退出app时结束所有Activity
         */
        fun finishAll() {
            for (activity in activities) {
                if (!activity.isFinishing) {
                    activity.finish()
                }
            }
        }

        /**
         * 判断是否存在Activity
         *
         * @param packageName 包名
         * @param className   activity全路径类名
         * @return `true`: 是<br></br>`false`: 否
         */
        fun isActivityExists(packageName: String, className: String): Boolean {
            val intent = Intent()
            intent.setClassName(packageName, className)
            return !(Utils.getContext().packageManager.resolveActivity(intent, 0) == null ||
                    intent.resolveActivity(Utils.getContext().packageManager) == null ||
                    Utils.getContext().packageManager.queryIntentActivities(intent, 0).size == 0)
        }

        /**
         * 获取launcher activity
         *
         * @param packageName 包名
         * @return launcher activity
         */
        fun getLauncherActivity(packageName: String): String {
            val intent = Intent(Intent.ACTION_MAIN, null)
            intent.addCategory(Intent.CATEGORY_LAUNCHER)
            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
            val pm = Utils.getContext().packageManager
            val info = pm.queryIntentActivities(intent, 0)
            for (aInfo in info) {
                if (aInfo.activityInfo.packageName == packageName) {
                    return aInfo.activityInfo.name
                }
            }
            return "no $packageName"
        }

        /**
         * 获取栈顶Activity
         *
         * @return 栈顶Activity
         */
//        val topActivity: Activity?
//            get() {
//                try {
//                    val activityThreadClass = Class.forName("android.app.ActivityThread")
//                    val activityThread =
//                        activityThreadClass.getMethod("currentActivityThread").invoke(null)
//                    val activitiesField = activityThreadClass.getDeclaredField("mActivities")
//                    activitiesField.isAccessible = true
//                    val activities: Map<*, *>
//                    if (Build.VERSION.SDK_INT < Build.VERSION_CODES.KITKAT) {
//                        activities = activitiesField.get(activityThread) as HashMap<*, *>
//                    } else {
//                        activities = activitiesField.get(activityThread) as ArrayMap<*, *>
//                    }
//                    for (activityRecord in activities.values) {
//                        val activityRecordClass = activityRecord.javaClass
//                        val pausedField = activityRecordClass.getDeclaredField("paused")
//                        pausedField.isAccessible = true
//                        if (!pausedField.getBoolean(activityRecord)) {
//                            val activityField = activityRecordClass.getDeclaredField("activity")
//                            activityField.isAccessible = true
//                            return activityField.get(activityRecord) as Activity
//                        }
//                    }
//                } catch (e: Exception) {
//                    e.printStackTrace()
//                }
//
//                return null
//            }
    }
}
